From 9ba378572e0427602b7619a30ea616395f3c1ed8 Mon Sep 17 00:00:00 2001
From: Aaradhana Sahu <quic_aarasahu@quicinc.com>
Date: Fri, 18 Nov 2022 17:32:19 +0530
Subject: [PATCH] cfg80211: get channel and txpower value for each link

Currently, wdev->valid_links set for multilink operation but channel
attribute values are filling when it's not set, because of this we
are not getting value of channel attribute for multilink.

So, added rdev_get_channel and rdev_get_tx_power function for each link.

Signed-off-by: Aaradhana Sahu <quic_aarasahu@quicinc.com>
---
 net/wireless/nl80211.c | 53 +++++++++++++++++++++++++++---------------
 1 file changed, 34 insertions(+), 19 deletions(-)

--- a/net/wireless/nl80211.c
+++ b/net/wireless/nl80211.c
@@ -4070,25 +4070,6 @@ static int nl80211_send_iface(struct sk_
 	    nla_put_u8(msg, NL80211_ATTR_4ADDR, wdev->use_4addr))
 		goto nla_put_failure;
 
-	if (rdev->ops->get_channel && !wdev->valid_links) {
-		struct cfg80211_chan_def chandef = {};
-		int ret;
-
-		ret = rdev_get_channel(rdev, wdev, 0, &chandef);
-		if (ret == 0 && nl80211_send_chandef(msg, &chandef))
-			goto nla_put_failure;
-	}
-
-	if (rdev->ops->get_tx_power) {
-		int dbm, ret;
-
-		ret = rdev_get_tx_power(rdev, wdev, &dbm);
-		if (ret == 0 &&
-		    nla_put_u32(msg, NL80211_ATTR_WIPHY_TX_POWER_LEVEL,
-				DBM_TO_MBM(dbm)))
-			goto nla_put_failure;
-	}
-
 	wdev_lock(wdev);
 	switch (wdev->iftype) {
 	case NL80211_IFTYPE_AP:
@@ -4150,10 +4131,37 @@ static int nl80211_send_iface(struct sk_
 			if (ret == 0 && nl80211_send_chandef(msg, &chandef))
 				goto nla_put_failure;
 
+			if (rdev->ops->get_tx_power) {
+				int dbm, ret;
+				/* TODO find tx_power according to link_id */
+				ret = rdev_get_tx_power(rdev, wdev, &dbm);
+				if (ret == 0 &&
+				    nla_put_u32(msg, NL80211_ATTR_WIPHY_TX_POWER_LEVEL,
+				    DBM_TO_MBM(dbm)))
+					goto nla_put_failure;
+			}
 			nla_nest_end(msg, link);
 		}
 
 		nla_nest_end(msg, links);
+	} else {
+		if (rdev->ops->get_channel) {
+			struct cfg80211_chan_def chandef = {};
+			int ret;
+
+			ret = rdev_get_channel(rdev, wdev, 0, &chandef);
+			if (ret == 0 && nl80211_send_chandef(msg, &chandef))
+				goto nla_put_failure;
+		}
+
+		if (rdev->ops->get_tx_power) {
+			int dbm, ret;
+			ret = rdev_get_tx_power(rdev, wdev, &dbm);
+			if (ret == 0 &&
+			    nla_put_u32(msg, NL80211_ATTR_WIPHY_TX_POWER_LEVEL,
+			    DBM_TO_MBM(dbm)))
+				goto nla_put_failure;
+		}
 	}
 
 	genlmsg_end(msg, hdr);
